components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/model/TokenRequest.java

```diff
@@ -0,0 +1,152 @@
+
+    static {
+        headersToAvoid.add("authorization");
+        headersToAvoid.add("cookie");
+        headersToAvoid.add("set-cookie");
+        headersToAvoid.add("accept-encoding");
+        headersToAvoid.add("accept-language");
+        headersToAvoid.add("content-length");
+        headersToAvoid.add("content-type");
+        // parameters from authorization code grant
+        paramsToAvoid.add("code");
+        paramsToAvoid.add("client_id");
+        paramsToAvoid.add("client_secret");
+        paramsToAvoid.add("redirect_uri");
+        paramsToAvoid.add("grant_type");
+        paramsToAvoid.add("scope");
+        // parameters from password grant
+        paramsToAvoid.add("username");
+        paramsToAvoid.add("password");
+        // parameters from refresh token grant
+        paramsToAvoid.add("refresh_token");
+    }
+
+    private final String clientId;
+    private final String grantType;
+    private final String redirectUri;
+    private final List<String> scopes;
+
+    private TokenRequest(Builder builder) {
+
+        this.clientId = builder.clientId;
+        this.grantType = builder.grantType;
+        this.redirectUri = builder.redirectUri;
+        this.scopes = builder.scopes;
+        this.additionalHeaders = builder.additionalHeaders;
+        this.additionalParams = builder.additionalParams;
+    }
+
+    public String getClientId() {
+
+        return clientId;
+    }
+
+    public String getGrantType() {
+
+        return grantType;
+    }
+
+    public String getRedirectUri() {
+
+        return redirectUri;
+    }
+
+    public List<String> getScopes() {
+
+        return scopes;
+    }
+
+    /**
+     * Builder for TokenRequest.
+     */
+    public static class Builder {
+
+        private final Map<String, String[]> additionalHeaders = new HashMap<>();
+        private final Map<String, String[]> additionalParams = new HashMap<>();
+        private String clientId;
+        private String grantType;
+        private String redirectUri;
+        private List<String> scopes = new ArrayList<>();
+
+        public Builder clientId(String clientId) {
+
+            this.clientId = clientId;
+            return this;
+        }
+
+        public Builder grantType(String grantType) {
+
+            this.grantType = grantType;
+            return this;
+        }
+
+        public Builder redirectUri(String redirectUri) {
+
+            this.redirectUri = redirectUri;
+            return this;
+        }
+
+        public Builder scopes(List<String> scopes) {
+
+            this.scopes = scopes;
+            return this;
+        }
+
+        public Builder addAdditionalHeader(String key, String[] value) {
+
+            if (!headersToAvoid.contains(key.toLowerCase())) {
+                this.additionalHeaders.put(key, value);
+            }
+            return this;
+        }
+
+        public Builder addAdditionalParam(String key, String[] value) {
+
+            if (!paramsToAvoid.contains(key)) {
```

Comment:
Let's check with out case sensitivity

Reply:
Well request parameters are case sensitive (Refer thread below that indicates such). So, IMO we should not treat is as case insensitive.

https://stackoverflow.com/questions/8969170/httpservletrequest-parameter-lowercase